### PR TITLE
add apollo client

### PR DESCRIPTION
**Pull request by gokhanap**

### CODE REVIEW

File: src/server.js [hunk 1]
```diff
@@ -85,6 +99,11 @@ server.get('/*', (req, res) => {
     <script>
       window.__PRELOADED_STATE__ = ${serialize(finalState)}
     </script>
+    <script>
+      window.__APOLLO_STATE__ = ${JSON.stringify(initialApolloState)
+    .replace(/</g, '\\u003c')}
```
Comment (SelmanKahya) [inline src/server.js]: why do we need to replace it?

File: src/server.js [hunk 2]
```diff
@@ -29,20 +32,31 @@ if (process.env.NODE_ENV !== 'test') {
...
+      <Provider store={store}>
+        <StaticRouter location={req.path} context={context}>
+          <App />
+        </StaticRouter>
+      </Provider>
+    </ApolloProvider>
   );
 
+  try {
+    await getDataFromTree(<Root />);
```
Comment (SelmanKahya) [inline src/server.js]: you are calling getDataFromTree but never store the return value? maybe client.extract automagically handles it?

File: src/components/pages/home.js [hunk 3]
```diff
@@ -3,15 +3,45 @@ import Header from '@/components/common/header';
...
+const TEST_GQL = gql`
+      {
+        rates(currency: "USD") {
+          currency
+        }
+      }
+    `;
+
+const TestComp = () => (
+  <Query query={TEST_GQL}>
```
Comment (SelmanKahya) [inline src/components/pages/home.js]: is this working as expected?

File: src/apollo.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable no-undef */
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable sort-keys */
```
Comment (SelmanKahya) [inline src/apollo.js]: lets not disable these rules

File: src/apollo.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable no-undef */
```
Comment (SelmanKahya) [inline src/apollo.js]: can we put it under `src/api/apollo.js`

File: src/apollo.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable sort-keys */
+import { InMemoryCache } from 'apollo-cache-inmemory';
+import { ApolloClient } from 'apollo-client';
+import { HttpLink } from 'apollo-link-http';
+import fetch from 'node-fetch';
+
+export default new ApolloClient({
+  connectToDevTools: process.browser,
```
Comment (SelmanKahya) [inline src/apollo.js]: can we only connectToDevTools if env === staging? https://github.com/jaredpalmer/razzle#environment-variables

File: src/apollo.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  ssrMode: !process.browser,
+  link: new HttpLink({
+    // uri: 'https://api.graphcms.com/simple/v1/cjhdcwrb98if90109o4pzawaq',
+    uri: 'https://w5xlvm3vzz.lp.gql.zone/graphql',
+    // uri: 'http://localhost:3000/gql',
+    fetch,
+  }),
+  cache: process.browser
+    ? new InMemoryCache().restore(window.__APOLLO_STATE__)
+    : new InMemoryCache(),
```
Comment (SelmanKahya) [inline src/apollo.js]: do you know how it caches on server side? will it invalidate the cache on mutations?

File: src/server.js (hunk 1 above)
Comment (gokhanap) [reply src/server.js]: not necessary. removed.

**Comment by gokhanap** [comment]
some test suites failed on these changes.
test suites will also require client.
added client and apollo provider into relevant test suites.
all tests passed.

File: src/apollo.js (hunk 6 above)
Comment (gokhanap) [reply src/apollo.js]:
```
  connectToDevTools: process.env.NODE_ENV === 'development',
```
there's no staging env on document. can we use development?

File: src/apollo.js (hunk 7 above)
Comment (gokhanap) [reply src/apollo.js]: cant figure out yet

File: src/api/apollo.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import { InMemoryCache } from 'apollo-cache-inmemory';
+import { ApolloClient } from 'apollo-client';
+import { HttpLink } from 'apollo-link-http';
+import fetch from 'node-fetch';
+
+export default new ApolloClient({
+  cache: process.browser
```
Comment (SelmanKahya) [inline src/api/apollo.js]: i dont think process.browser would work, we can use utils.js - isServer

File: src/api/apollo.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  // eslint-disable-next-line
+    ? new InMemoryCache().restore(window.__APOLLO_STATE__)
+    : new InMemoryCache(),
+  connectToDevTools: process.env.NODE_ENV === 'development',
+  link: new HttpLink({
+    fetch,
+    uri: 'https://api.graph.cool/simple/v1/swapi',
+    // uri: 'http://localhost:3000/gql',
+  }),
+  ssrMode: !process.browser,
```
Comment (SelmanKahya) [inline src/api/apollo.js]: same

**Pull request merged**